packages/debug/src/browser/debug-contribution-manager.ts

```diff
@@ -0,0 +1,247 @@
+
+    async resolveDebugConfiguration(config: DebugConfiguration, workspaceFolderUri: string | undefined): Promise<DebugConfiguration> {
+        const contributor = this.pluginContributors.get(config.type);
+        if (contributor) {
+            const resolved = await contributor.resolveDebugConfiguration(config, workspaceFolderUri);
+            return resolved || config;
+        } else {
+            return this.debugService.resolveDebugConfiguration(config, workspaceFolderUri);
+        }
+    }
+
+    async getDebuggersForLanguage(language: string): Promise<DebuggerDescription[]> {
+        const debuggers = await this.debugService.getDebuggersForLanguage(language);
+
+        for (const contributor of this.pluginContributors.values()) {
+            const languages = await contributor.getSupportedLanguages();
+            if (languages && languages.indexOf(language) !== -1) {
+                debuggers.push(contributor.description);
+            }
+        }
+
+        return debuggers;
+    }
+
+    async getSchemaAttributes(debugType: string): Promise<IJSONSchema[]> {
+        const contributor = this.pluginContributors.get(debugType);
+        if (contributor) {
+            return contributor.getSchemaAttributes();
+        } else {
+            return this.debugService.getSchemaAttributes(debugType);
+        }
+    }
+
+    async getConfigurationSnippets(): Promise<IJSONSchemaSnippet[]> {
+        let snippets = await this.debugService.getConfigurationSnippets();
+
+        for (const contributor of this.pluginContributors.values()) {
+            const contribSnippets = await contributor.getConfigurationSnippets();
+            if (contribSnippets) {
+                snippets = snippets.concat(contribSnippets);
+            }
+        }
+
+        return snippets;
+    }
+
+    async create(config: DebugConfiguration): Promise<string> {
+        const contributor = this.pluginContributors.get(config.type);
+        if (contributor) {
+            return contributor.createDebugSession(config);
+        } else {
+            return this.debugService.createDebugSession(config);
+        }
+    }
+
+    async stop(debugType: string, sessionId: string): Promise<void> {
+        const contributor = this.pluginContributors.get(debugType);
+        if (contributor) {
+            return contributor.terminateDebugSession(sessionId);
+        } else {
+            return this.debugService.terminateDebugSession(sessionId);
+        }
+    }
+
+    private async isContributorRegistered(debugType: string): Promise<boolean> {
+        const registeredTypes = await this.debugTypes();
+        return registeredTypes.indexOf(debugType) !== -1;
+    }
+}
+
+/**
+ * Session factory for a client debug session that communicates with debug adapter contributed as plugin.
+ * The main difference is to use a connection factory that creates [IWebSocket](#IWebSocket) over Rpc channel.
+ */
+class PluginDebugSessionFactory implements DebugSessionFactory {
```

Comment:
a class should go to its own file, a file name should be  a class name

we try not put several classes in one file with a randome name, also we try to avoid classes with the same name in the repo

it is easy navigation